src/business-logic/services/database-change-ingest-service.ts

```diff
@@ -0,0 +1,283 @@
+        rundownChangeListener,
+        segmentChangedListener,
+        partChangedListener
+      )
+    }
+    return this.instance
+  }
+
+  private readonly eventQueue: (() => Promise<void>)[] = []
+  private isExecutingEvent: boolean = false
+  private lastBulkExecutionStartTimestamp: number = 0
+  private readonly rundownIdsToBuild: Set<string> = new Set<string>()
+
+  private constructor(
+    private readonly rundownRepository: RundownRepository,
+    private readonly timelineRepository: TimelineRepository,
+    private readonly timelineBuilder: TimelineBuilder,
+    private readonly eventEmitter: RundownEventEmitter,
+    private readonly eventBuilder: RundownEventBuilder,
+    rundownChangeListener: DataChangedListener<Rundown>,
+    segmentChangedListener: DataChangedListener<Segment>,
+    partChangedListener: DataChangedListener<Part>
+  ) {
+    this.listenForRundownChanges(rundownChangeListener)
+    this.listenForSegmentChanges(segmentChangedListener)
+    this.listenForPartChanges(partChangedListener)
+  }
+
+  private listenForRundownChanges(rundownChangeListener: DataChangedListener<Rundown>): void {
+    rundownChangeListener.onCreated(rundown => this.enqueueEvent(() => this.createRundown(rundown)))
+    rundownChangeListener.onUpdated(rundown => this.enqueueEvent(() => this.updateRundown(rundown)))
+    rundownChangeListener.onDeleted(rundownId => this.enqueueEvent(() => this.deleteRundown(rundownId)))
+  }
+
+  private listenForSegmentChanges(segmentChangedListener: DataChangedListener<Segment>): void {
+    segmentChangedListener.onCreated(segment => this.enqueueEvent(() => this.createSegment(segment)))
+    segmentChangedListener.onUpdated(segment => this.enqueueEvent(() => this.updatedSegment(segment)))
+    segmentChangedListener.onDeleted(segmentId => this.enqueueEvent(() => this.deleteSegment(segmentId)))
+  }
+
+  private listenForPartChanges(partChangedListener: DataChangedListener<Part>): void {
+    partChangedListener.onCreated(part => this.enqueueEvent(() => this.createPart(part)))
+    partChangedListener.onUpdated(part => this.enqueueEvent(() => this.updatePart(part)))
+    partChangedListener.onDeleted(partId => this.enqueueEvent(() => this.deletePart(partId)))
+  }
+
+  private enqueueEvent(event: () => Promise<void>): void {
+    this.eventQueue.push(event)
+    this.executeNextEvent()
+  }
+
+  private executeNextEvent(): void {
+    if (this.isExecutingEvent) {
+      return
+    }
+    if (Date.now() - this.lastBulkExecutionStartTimestamp >= BULK_EXECUTION_TIMESPAN_IN_MS) {
+      this.lastBulkExecutionStartTimestamp = Date.now()
+    }
+    const eventCallback: (() => Promise<void>) | undefined = this.eventQueue.shift()
+    if (!eventCallback) {
+      return
+    }
+
+    this.isExecutingEvent = true
+    eventCallback()
+      .catch(error => console.error('Error when executing Ingest event:', error))
+      .finally(() => {
+        this.isExecutingEvent = false
+        void this.buildRundowns()
```

Comment:
What do you suggest?

Reply:
Either of the two snippets below. What I want is that if buildRundowns fails, we catch the error. I know that currently that shouldn't happen, but we will most likely forget to update this, if we change the internals of buildRundowns.

```typescript
eventCallback()
  .catch(error => ...)
  .then(() => this.buildRundowns())
  .catch(error => console.log('Failed building rundowns after event execution', error))
  .finally(() => {
    this.isExecutingEvent = false
    this.executeNextEvent()
  })
```
or
```typescript
eventCallback()
  .catch(error => ...)
  .finally(() => {
    this.isExecutingEvent = false
    this.buildRundowns().catch(error => console.error('Failed building rundowns after event execution', error))
    this.executeNextEvent()
  })
```

Reply:
So I assume you then also wants to remove the try-catch from `buildRundowns` because I am handling the error there? 
![image](https://github.com/tv2/sofie-server/assets/23049647/423b0fde-02f2-47fa-8268-e8aa09a3d752)
If we don't remove it there then we just handle the same error twice (meaning that the second catch is redundant since it will never be called)